### PR TITLE
Add Location Strategy Keyword

### PR DESCRIPTION
**Pull request by just-be-dev**
This was discussed both in #333 and #364 and here's a preliminary thought that I had. 

Instead of tying the `Add Location Strategy` to javascript as I originally intended I've designed it so that a location strategy is actually implemented as a keyword. This gives a lot of flexibility in that location strategies can (if desired) be written directly in RF as a user keyword.

Keep in mind that this is just a rough draft to get started with. 

Idealistically I'd have this

``` robotframework
*** Settings ***
Library    Selenium2Library
Resource    Locators
Resource    Web Navigation Stuff

*** Keyword ***
Setup Locators
    Add Location Strategy    custom    My Custom Locator

*** Test Cases ***
My Test
    [Setup]    Setup Locators
    Open Browser To My Site
    Click Element    custom=mySelector[example=true]
```

### CODE REVIEW

**Comment by just-be-dev** [comment]
@emanlove, could you take a look at this and give me your thoughts?

Thanks.

**Comment by just-be-dev** [comment]
This is now feature complete and ready for review. I've added tests and documentation. 

It includes two new keywords:
- `Add Locator Strategy`
- `Remove Locator Strategy`

The locator strategy itself is a user keyword as shown in the documentation. This allows easy integration with JavaScript (which was my original goal), but also covers any other applicable situation. 

An added locator strategy will stay present throughout the duration of testing unless removed by `Remove Locator Strategy`.

**Comment by just-be-dev** [comment]
`Add Locator Strategy` is a global action. When this keyword is called, the locator strategy that is specified doesn't go away unless Remove Locator Strategy is subsequently called. 

Global alterations of a testing environment by a test are generally unwise. @pekkaklarck, @emanlove what do you think of this behavior? I suppose it would be possible through library listener methods to automatically unregister the locator strategy after the scope it was registered in has subsided. 

Which would be better? The current behavior (after it's set anywhere it stays until it is removed) or the scoped behavior (where if it's set in a test it's removed after that test is complete).

**Comment by pekkaklarck** [comment]
@zephraph, automatically removing locator strategies sounds like a good idea. Library listeners ought to make implementing it pretty simple. Should probably be configurable with an argument like `auto_remove=False`, though.

**Comment by just-be-dev** [comment]
I've added a specific test because I was in talks with @IlfirinPL on #376. This test case will likely be removed before merge along with the changes mentioned by @pekkaklarck above. 

Don't merge this as is.

**Comment by just-be-dev** [comment]
@pekkaklarck, when you get a chance could you review this?

Adding the auto unregister made it a bit more complicated.

**Comment by just-be-dev** [comment]
@emanlove, could you please review this? Actually, look over everything in the 1.7 Milestone. I really want to go ahead and push a new release this weekend.

**Comment by just-be-dev** [comment]
A large part of the bulk of this is the scoping as suggested by @pekkaklarck. The original implementation made it persistent by default. This implementation makes it scoped by default (which required hooking into listeners). I try to make a little event system so if we needed to do stuff like this in the future we could just add new events.

**Comment by pekkaklarck** [comment]
Sorry, I was on a business trip with limited time for other work. What commits/code should be reviewed?

**Comment by just-be-dev** [comment]
Just look [here](https://github.com/rtomac/robotframework-selenium2library/pull/365/files) and see if anything jumps out at you. I really would like you to look over the event system I wrote for scoping the  locator registration and tell me if it's too much.

File: src/Selenium2Library/locators/elementfinder.py [hunk 1]
```diff
@@ -35,6 +37,30 @@ def find(self, browser, locator, tag=None):
         (tag, constraints) = self._get_tag_and_constraints(tag)
         return strategy(browser, criteria, tag, constraints)
 
+    def register(self, strategy, persist):
+        if(strategy.name in self._strategies):
```
Comment (pekkaklarck) [inline src/Selenium2Library/locators/elementfinder.py]: Parens? Also in `unregister` below. #pep8

File: src/Selenium2Library/utils/events/__init__.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+_events = []
+
+__all__ = [
+	"on",
+	"dispatch",
+	"register_event"
+]
+
+def on(event_name, *args, **kwargs):
+	for event in _registered_events:
```
Comment (pekkaklarck) [inline src/Selenium2Library/utils/events/__init__.py]: Are tabs used here or is indentation this long?

File: src/Selenium2Library/utils/librarylistener.py [hunk 3]
```diff
@@ -0,0 +1,18 @@
+import events as event
+from robot.api import logger
+
+class LibraryListener(object):
+
+	ROBOT_LISTENER_API_VERSION = 2
+
+	def _start_suite(self, name, attrs):
```
Comment (pekkaklarck) [inline src/Selenium2Library/utils/librarylistener.py]: It would be OK not to have the leading underscore because this class is not used as a library.

**Comment by pekkaklarck** [comment]
The event system looks pretty neat but I'm not entirely sure is it necessary for this usage. Are you planning to use it for something else too? If not, then there's a risk that YAGNI.

I also added some nitpicking comments about style. I'm not sure is there any agreed styles for this project so feel free to ignore them.

File: src/Selenium2Library/locators/elementfinder.py (hunk 1 above)
Comment (just-be-dev) [inline src/Selenium2Library/locators/elementfinder.py]: Ah, good catch. That's my Javascript habits kicking in.

File: src/Selenium2Library/utils/events/__init__.py (hunk 2 above)
Comment (just-be-dev) [inline src/Selenium2Library/utils/events/__init__.py]: Ah, crap. I had didn't have my editor set on soft-tabs. Good catch.

**Comment by just-be-dev** [comment]
Well, it's quite possible I could use it again in the future, but I can't guarantee that. That's just how I thought to implement it. Also, I trimmed the scope events down a bit. It's more DRY now.

**Comment by pekkaklarck** [comment]
Ok, looks good.

**Comment by just-be-dev** [comment]
So, are we good to merge?

**Comment by divfor** [comment]
I have  an idea on similar enhancement:
https://github.com/divfor/robotframework-selenium2library/pull/6
This enhancement is to simplify locator usage. Generally it is difficult for junior engineers and customer to understand the locators written in user keywords or test steps. It would be better if they can use Web UI texts/words prompting for target element instead of locators. The real locators could be mapped by prompting UI text in variable files created and maintained by senior engineers. If so, the test cases will be more easier to read.

**Comment by just-be-dev** [comment]
I'm not really following you. The custom locators written here should be specified in a resource file or library, not included in a test. I wrote this feature with javascript frameworks like angular and ExtJs in mind, though it's not limited to those sorts of things. I would say it's more up to the locator developer to document their locator usage. By and large this should be a rarely used feature though.

**Pull request merged**